Service/src/main/java/org/apidb/apicommon/service/services/jbrowse/JBrowseUserDatasetsService.java

```diff
@@ -1,161 +1,206 @@
-          Long datasetId = dataset.getUserDatasetId();
-          UserDatasetMeta datasetMeta = dataset.getMeta();
-          String datasetName = datasetMeta.getName();
-          String datasetSummary = datasetMeta.getSummary();
-
-          for (UserDatasetFile file : dataset.getFiles().values()) {
-              String fileName = file.getFileName();
-              if(!fileName.toUpperCase().endsWith(".BW"))
-                  continue ;
-
-              String urlTemplate = "/a/service/users/current/user-datasets/" + datasetId + "/user-datafiles/" + fileName;
-
-              JSONObject json = new JSONObject();
-              json.put("storeClass", "JBrowse/Store/SeqFeature/BigWig");
-              json.put("urlTemplate", urlTemplate);
-              json.put("yScalePosition",  "left");
-              json.put("key", datasetName + " " + fileName);
-              json.put("label", datasetName + " " + fileName);
-              json.put("type", "JBrowse/View/Track/Wiggle/XYPlot");
-              json.put("category", "My Data from Galaxy");
-              json.put("min_score", 0);
-              json.put("max_score", maxScore);
-
-              JSONObject style = new JSONObject();
-              style.put("pos_color", "#5B2C6F");
-
-              JSONObject metadata = new JSONObject();
-              metadata.put("subcategory", subcategory);
-              metadata.put("dataset", datasetName);
-              metadata.put("trackType", "Coverage");
-              metadata.put("mdescription", datasetSummary);
-
-              json.put("metadata", metadata);
-              json.put("style", style);
-
-              samplesJson.put(json);
+  /**
+   * Fetch all tracks from the filesystem for a given dataset reference. If the dataset has no installed files in this
+   * project and empty list is returned.
+   */
+  private List<JBrowseTrack> fetchTracksFromFilesystem(VDIDatasetReference vdiDatasetReference) {
+    final String vdiDatasetsDir = getWdkModel().getProperties().get(VDI_DATASET_DIR_KEY);
+    final String buildNumber = getWdkModel().getBuildNumber();
+    final java.nio.file.Path datasetDir = Paths.get(vdiDatasetsDir, "build-" + buildNumber, getWdkModel().getProjectId(), vdiDatasetReference.getId());
+    return Arrays.stream(Optional.ofNullable(datasetDir.toFile().listFiles()).orElse(new File[0]))
+        .map(bwFile -> {
+          final JBrowseTrack jBrowseTrack = new JBrowseTrack();
+
+          jBrowseTrack.setKey(vdiDatasetReference.getName() + " " + bwFile.getName());
+          jBrowseTrack.setLabel(vdiDatasetReference.getName() + " " + bwFile.getName());
+          jBrowseTrack.setUrlTemplate(String.format("/a/service/users/current/user-datasets-jbrowse/data?datasetID=%s?data=%s",
+              vdiDatasetReference.getId(), bwFile.getName()));
+
+          JBrowseTrack.Metadata metadata = new JBrowseTrack.Metadata();
+          if (vdiDatasetReference.getType().equalsIgnoreCase(VDIDatasetType.BIGWIG.getVdiName())) {
```

Comment:
Feel like this comparison/lookup could be done in the enum (find the subcatetory), then just set returned value onto the track and in metadata.

Reply:
Fixed with the enum change above